src/resource/SimpleResource.ts

```diff
@@ -0,0 +1,366 @@
+import { memoize } from 'lodash';
+import { AbstractInstanceType, Method, RequestOptions } from '~/types';
+
+import { ReadShape, MutateShape, DeleteShape } from './types';
+import { schemas, SchemaDetail, SchemaList } from './normal';
+
+const getEntitySchema: <T extends typeof SimpleResource>(
+  M: T,
+) => schemas.Entity<AbstractInstanceType<T>> = memoize(
+  <T extends typeof SimpleResource>(M: T) => {
+    const e = new schemas.Entity(
+      M.getKey(),
+      {},
+      {
+        idAttribute: (value, parent, key) => {
+          const id = M.pk(value) || key;
+          if (process.env.NODE_ENV !== 'production' && id === null) {
+            throw new Error(
+              `Missing usable resource key when normalizing response.
+
+This is likely due to a malformed response.
+Try inspecting the network response or fetch() return value.
+`,
+            );
+          }
+          return id.toString();
+        },
+        processStrategy: value => {
+          return M.fromJS(value);
+        },
+        mergeStrategy: (
+          a: AbstractInstanceType<T>,
+          b: AbstractInstanceType<T>,
+        ) => (a.constructor as T).merge(a, b),
+      },
+    );
+    // TODO: long term figure out a plan to actually denormalize
+    (e as any).denormalize = function denormalize(entity: any) {
+      return entity;
+    };
+    return e;
+  },
+) as any;
+
+const DefinedMembersKey = Symbol('Defined Members');
+type Filter<T, U> = T extends U ? T : never;
+interface SimpleResourceMembers<T extends typeof SimpleResource> {
+  [DefinedMembersKey]: (Filter<keyof AbstractInstanceType<T>, string>)[];
+}
+
+/** Represents an entity to be retrieved from a server. Typically 1:1 with a url endpoint. */
+export default abstract class SimpleResource {
+  // typescript todo: require subclasses to implement
+  /** Used as base of url construction */
+  static readonly urlRoot: string;
+  /** A unique identifier for this SimpleResource */
+  abstract pk(): string | number | null;
+
+  /** SimpleResource factory. Takes an object of properties to assign to SimpleResource. */
+  static fromJS<T extends typeof SimpleResource>(
+    this: T,
+    props: Partial<AbstractInstanceType<T>>,
+  ) {
+    // we type guarded abstract case above, so ok to force typescript to allow constructor call
+    const instance = new (this as any)(props) as AbstractInstanceType<T>;
+
+    if (instance.pk === undefined)
```

Comment:
except this part was changed. This is better detection because there can be multiple abstract classes in the hierarchy beyond this.